consensus/cryptarchia-sync/src/behaviour.rs

```diff
@@ -0,0 +1,489 @@
+    incoming_streams: IncomingStreams,
+    /// List of connected peers.
+    peers: Vec<PeerId>,
+    /// Futures for managing the progress of locally initiated block downloads.
+    locally_initiated_downloads: FuturesUnordered<BoxFuture<'static, Result<(), ChainSyncError>>>,
+    /// Futures for managing the progress of externally initiated block
+    /// downloads.
+    externally_initiated_downloads:
+        FuturesUnordered<BoxFuture<'static, Result<(), ChainSyncError>>>,
+    /// Futures for reading incoming download requests. After the request is
+    /// read, sending blocks is handled by `externally_initiated_downloads`.
+    external_pending_download_requests: FuturesUnordered<
+        BoxFuture<'static, Result<(Stream, DownloadBlocksRequest), ChainSyncError>>,
+    >,
+    /// Futures for closing incoming streams that were rejected due to excess
+    incoming_streams_to_close: FuturesUnordered<BoxFuture<'static, ()>>,
+}
+
+impl Default for Behaviour {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Behaviour {
+    #[must_use]
+    pub fn new() -> Self {
+        let stream_behaviour = StreamBehaviour::new();
+        let mut control = stream_behaviour.new_control();
+        let incoming_streams = control
+            .accept(SYNC_PROTOCOL)
+            .expect("Failed to accept incoming streams for sync protocol");
+        Self {
+            stream_behaviour,
+            control,
+            incoming_streams,
+            peers: Vec::new(),
+            locally_initiated_downloads: FuturesUnordered::new(),
+            externally_initiated_downloads: FuturesUnordered::new(),
+            external_pending_download_requests: FuturesUnordered::new(),
+            incoming_streams_to_close: FuturesUnordered::new(),
+        }
+    }
+
+    fn add_peer(&mut self, peer: PeerId) {
+        if !self.peers.contains(&peer) {
+            self.peers.push(peer);
+        }
+    }
+
+    fn remove_peer(&mut self, peer: &PeerId) {
+        self.peers.retain(|p| p != peer);
+    }
+
+    fn choose_peer(&self) -> Option<PeerId> {
+        self.peers.clone().into_iter().choose(&mut rand::rng())
+    }
+
+    pub fn start_blocks_download(
+        &mut self,
+        target_block: Option<HeaderId>,
+        local_tip: HeaderId,
+        immutable_block: HeaderId,
+        additional_blocks: Vec<HeaderId>,
+        rx_blocks: mpsc::Sender<BlocksResponse>,
+    ) -> Result<(), ChainSyncError> {
+        let peer_id = self.choose_peer().ok_or_else(|| {
+            ChainSyncError::StartSyncError("No peers available for chain sync".into())
+        })?;
+
+        let request =
+            DownloadBlocksRequest::new(target_block, local_tip, immutable_block, additional_blocks);
+        let mut task = DownloadBlocksTask::new(request, rx_blocks);
+
+        let control = self.control.clone();
+        self.locally_initiated_downloads
+            .push(async move { task.download_blocks(peer_id, control).await }.boxed());
```

Comment:
Yeah, this would probably just go away. We do not need to poll from futures here, just return the computation to the one that is gonna actively use it.